### ISSUE TITLE
Poor performance in MPI code after OMP_PROC_BIND environment variable set as per Kokkos::OpenMP::initialize warning 

### ISSUE THREAD

**Issue by Chrismarsh** [kokkos/kokkos#6180]
**Describe the bug**
I have code that uses Trillinos/Kokkos in a MPI-OpenMP hybrid manner. There is a warning from Kokkos that states:

```
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
```

My feedback is that this is not the best warning as it can lead to significant performance degradation in a hybrid OMP-MPI parallel.

I had a user of my code see this warning and set those env vars as directed.  However, when these are set, it caused all the MPI ranks to run on a single core, leaving the rest of the cores unused. Of note is that there was no warning from MPI that an over subscription occurred. It seemed that when the OMP threads were created, their parent process probably re-pinned themselves according to these OMP env vars. As each OMP-enabled process is unaware of the others, they all tried to pin to the core at offset 0. 

Resolving the above was tedious as it isn't clear that setting OMP vars should affect the behaviour of MPI like this. We verified all the common culprits like MPI bindings, etc. But it was finally tracked down to setting either of these env vars. 

Therefore, my suggestion is to remove this as a warning. Alternatively, caveat it with "in an MPI hybrid context this can cause unforeseen issues".

**Please also include the following items to support reproducing the bug**
1. gcc 9.3.0
2. Whatever version of Kokkos that is in Trilinos 13.4.0. I believe it is commit `f2a05d316596ac8a06fb4582bc2bc423033e4396` as per [here](https://github.com/trilinos/Trilinos/tree/trilinos-release-13-4-0/packages/kokkos).
3. Linux, omp4

**Comment by dalg24** [comment]
> Of note is that there was no warning from MPI that an over subscription occurred.

Can you elaborate I do not understand?

**Comment by Chrismarsh** [comment]
Oversubscribing happens when there are more ranks on a host than available compute slots. E.g., a fatal error is thrown by the MPI runtime if one did `mpirun -np 8 ./a.out` on a quad-core machine resulting in 2 ranks per core. 

Re-reading the documentation for oversubscribed suggests that I may be wrong and that 4 ranks on a quad-core system pinned to the same core might not throw this warning/fatal error.

**Comment by dalg24** [comment]
I should have been clearer.  I know what oversubscription means.  What I don't understand is the "no warning".  I expect Kokkos raised warnings if we detect that (number of MPI ranks on the local node) times (number of OpenMP threads) exceeds the hardware concurrency.

**Comment by dalg24** [comment]
I expect your user got something like
```
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false

...

Kokkos::OpenMP::initialize WARNING: You are likely oversubscribing your CPU cores.
                                    Detected: 10 cores per node.
                                    Detected: 8 MPI_ranks per node.
                                    Requested: 10 threads per process.
...

**Comment by dalg24** [comment]
Are you setting `OMP_NUM_THREADS` or telling Kokkos how many threads to use by some other way (e.g. `--kokkos-num-threads=???`)?  By default, if you don't do specify anything, we use them all.

Please review #6185

**Comment by crtrott** [comment]
I think the problem comes from MPI run in an unpinned manner. You need to tell mpi to use process masks so each process only sets a subset. My guess is that every rank sees all the cores. Not setting the omp variables just means all threads are free floating and you rely on os distributing threads. It will also mean you loose memory affinity every time the os decides to move your threads.

**Comment by crtrott** [comment]
So just to make this clear I think the warning is still correct. We could amend the warning to mention additional MPI concerns:
"When using MPI together with OpenMP, MPI processes must be bound to exclusive CPU masks for OpenMP binding to work as intended."

**Comment by Chrismarsh** [comment]
Sorry for the slow reply.

@dalg24, `OMP_NUM_THREADS` was set to `1`. 
@crtrott if we bound mpi via `--bind-to core --map-by core` then indeed the problem didn't occur.

I think the additional information in #6185  looks good! Thanks very much!

**Comment by crtrott** [comment]
You are welcome, and thanks for the usability report.

**Comment by lindsayad** [comment]
Hi we encountered a rough time with performance after setting the environment variables suggested by this warning. I am using mpich 4.3.1 and I do not see the mpi warning that was added in #6185. Stepping through the code indeed
```
(gdb) p env_var
$4 = 0x7fffc8aed09e "MPI_LOCALRANKID"
(gdb) p str
$5 = 0x0
```
I see `MPI_LOCALRANKID` published in a `launch_procs` static function in the hydra process manager source code, but I cannot break on that function; my best guess is that function is not being called. I do see in my `config.log` for my mpich build that hydra is my process manager.

Anyway I think that relying on this env variable is probably brittle for determining whether MPI is present
